### DISCUSSION TITLE
Residual computation

### DISCUSSION THREAD

**Discussion by MartinKarp** [Nek5000/nekRS#280] [General]
Hi all!

Thank you for you interesting paper "NekRS, a GPU-Accelerated Spectral Element Navier-Stokes Solver", really impressive runs!

One thing I wonder overall though is the regarding the residual in NekRS vs what has been used in Nek5000. From what I have understood in Nek500 the norms used are the L2 norm for the pressure and then the L2 norm weighted with inverted mass matrix for the velocity, both residuals scaled with the volume of the domain. I see now that L2 is used for both the pressure and velocity in NekRS and wonder why this has been changed. I also wonder if this is the reason why the tolerances in the paper are quite large i.e. 1e-4 for the pressure and 1e-6 for the velocity. From my experience with Nek5000 we have often run with tolerances around 1e-8 - 1e-10.

Very interested in your thinking on this and how one should think about the residuals/tolerances when running a simulation.

Best regards
Martin

**Comment by stgeke** [comment]
All linear solvers in nekRS are using the 2-norm of the (absolute or relative) residual similar to GMRE in Nek5000. You're right the CG solver uses different weights (to work properly in a dimensional context). This is inconsistent and should be fixed in Nek5000. 

I consider 1e-4/1e-6 as a good starting point for pressure/velocity - at least for the Pn/Pn formulation. For most engineering type cases this is typically accurate enough as time and/or spatial discretization errors will dominate. However that's something you want to verify/adjust for the case you're interested in.

**Reply by MartinKarp** [reply]
Thank you for your answer.

OK, so relative tolerance is supported! If one would like to run with relative tolerance, how does one activate it?  

I also wonder regarding the normalization of the residual with `norm_fac`  in Nek5000 and `elliptic->resNormFactor` in NekRS. I see it corresponds to dividing with the volume for each field we solve for (`elliptic->resNormFactor = 1 / (elliptic->Nfields * mesh->volume)` in `ellipticSolve.cpp`), but I wonder why we do this normalization and not use the L2/relative L2 norm as is?
